controllers/upgrade.go

```diff
@@ -777,57 +840,167 @@ func (r *RollingUpgradeContext) ClusterBallooning(batchSize int) (bool, int) {
 func (r *RollingUpgradeContext) CordonUncordonAllNodes(cordonNode bool) (bool, error) {
 	scalingGroup := awsprovider.SelectScalingGroup(r.RollingUpgrade.ScalingGroupName(), r.Cloud.ScalingGroups)
-	var instanceIDs []string
-	var err error
 
-	if cordonNode {
-		instanceIDs, err = r.Cloud.AmazonClientSet.DescribeInstancesWithoutTagValue(instanceStateTagKey, earlyCordonedTagValue)
-		if err != nil {
-			r.Error(err, "failed to describe instances for early-cordoning", "name", r.RollingUpgrade.NamespacedName())
-			return false, errors.Wrap(err, "failed to describe instances for early-cordoning")
-		}
-	} else {
-		instanceIDs, err = r.Auth.DescribeTaggedInstanceIDs(instanceStateTagKey, earlyCordonedTagValue)
-		if err != nil {
-			r.Info("failed to discover ec2 instances with early-cordoned tag", "name", r.RollingUpgrade.NamespacedName())
+	var processedCount, errorCount int
+
+	for _, instance := range scalingGroup.Instances {
+		// Skip terminating instances
+		if common.ContainsEqualFold(awsprovider.TerminatingInstanceStates, aws.StringValue(instance.LifecycleState)) {
+			continue
 		}
 
-		r.Info("removing early-cordoning tag while uncordoning instances", "name", r.RollingUpgrade.NamespacedName())
-		if err := r.Auth.UntagEC2instances(instanceIDs, instanceStateTagKey, earlyCordonedTagValue); err != nil {
-			r.Info("failed to delete early-cordoned tag for instances", "name", r.RollingUpgrade.NamespacedName())
+		// Skip non-drifted instances
+		if !r.IsInstanceDrifted(instance) {
+			continue
 		}
-		// add unit test as well.
 
-	}
+		node := kubeprovider.SelectNodeByInstanceID(*instance.InstanceId, r.Cloud.ClusterNodes)
+		if node == nil {
+			r.Info("node object not found in clusterNodes, unable to process node", "instanceID", instance.InstanceId, "name", r.RollingUpgrade.NamespacedName())
+			continue
+		}
 
-	for _, instanceID := range instanceIDs {
-		if instance := awsprovider.SelectScalingGroupInstance(instanceID, scalingGroup); !reflect.DeepEqual(instance, &autoscaling.Instance{}) {
-			//Don't consider if the instance is in terminating state.
-			if !common.ContainsEqualFold(awsprovider.TerminatingInstanceStates, aws.StringValue(instance.LifecycleState)) {
-				node := kubeprovider.SelectNodeByInstanceID(*instance.InstanceId, r.Cloud.ClusterNodes)
-				if node == nil {
-					r.Info("node object not found in clusterNodes, unable to early-cordon node", "instanceID", instance.InstanceId, "name", r.RollingUpgrade.NamespacedName())
+		if cordonNode {
+			// Skip already cordoned nodes to avoid unnecessary API calls
+			if node.Spec.Unschedulable {
+				// Check if it was cordoned by upgrade-manager
+				if node.Annotations != nil && node.Annotations[EarlyCordonAnnotationKey] == EarlyCordonAnnotationValue {
+					r.Info("node already cordoned by upgrade-manager, skipping", "instanceID", instance.InstanceId, "nodeName", node.Name, "name", r.RollingUpgrade.NamespacedName())
```

Comment:
Remove the "... by upgrade-manager" from the log. Node could be cordoned by other entities too - cluster-autoscaler, lifecycle-manager, node-reaper.